File: src/main/scala/dregex/impl/Normalization.scala
package dregex.impl

import scala.collection.JavaConverters._

sealed trait Normalization {
  def normalize(s: CharSequence): CharSequence
}

object Normalization {

  case object NoNormalization extends Normalization {
    def normalize(s: CharSequence) = s
  }

  case object LowerCase extends Normalization {
    def normalize(s: CharSequence): String = {
      val builder = new StringBuilder
      for (i <- 0 until s.length) {
        builder.append(AsciiHelper.toLower(s.charAt(i)))
      }
      builder.toString
    }
  }

  case object UnicodeLowerCase extends Normalization {
    def normalize(s: CharSequence): String = {
      val builder = new StringBuilder
      for (codePoint <- s.codePoints.iterator.asScala) {
        builder.append(Character.toLowerCase(codePoint))
      }
      builder.toString
    }
  }

}

File: src/main/java/dregex/impl/Normalization.java
package dregex.impl;

public enum Normalization implements Normalizer {
    NoNormalization {
        @Override
        public CharSequence normalize(CharSequence str) {
            return str;
        }
    },

    LowerCase {
        @Override
        public CharSequence normalize(CharSequence str) {
            var builder = new StringBuilder(str.length());
            for (int i = 0; i < str.length(); i++) {
                char ch = str.charAt(i);
                boolean isAsciiUpperCase = ch >= 'A' && ch <= 'Z';
                // cast to prevent promotion to int and calling the wrong overload
                builder.append(isAsciiUpperCase ? ((char) (ch + 0x20)) : ch);
            }
            return builder.toString();
        }
    },

    UnicodeLowerCase {
        @Override
        public CharSequence normalize(CharSequence str) {
            var builder = new StringBuilder();
            str.codePoints().forEach(codePoint -> {
                // cast to prevent calling the wrong overload
                builder.append((char) Character.toLowerCase(codePoint));
            });
            return builder.toString();
        }
    },

    CanonicalDecomposition {
        @Override
        public CharSequence normalize(CharSequence str) {
            return java.text.Normalizer.normalize(str, java.text.Normalizer.Form.NFD);
        }
    };
}
